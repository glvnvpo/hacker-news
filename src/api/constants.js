// @flow

type ID = number | string;

const BASE = 'https://hacker-news.firebaseio.com/v0';

export const NEW_STORIES = () => `${BASE}/newstories.json`;
export const ITEM = (id: ID)=> `${BASE}/item/${id}.json`;