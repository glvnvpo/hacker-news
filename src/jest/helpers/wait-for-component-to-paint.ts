import {act} from 'react-dom/test-utils';

export const waitForComponentToPaint = async () => {
	await act(async () => {
		await new Promise(resolve => setTimeout(resolve, 0));
	});
};