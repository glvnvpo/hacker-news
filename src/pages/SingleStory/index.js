// @flow

export {SingleStory} from './SingleStory';