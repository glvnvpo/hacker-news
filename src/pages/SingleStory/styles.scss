@import "~bootstrap/scss/bootstrap";
@import "../../style/variables";

.single-story {
  display: flex;
  justify-content: center;

  .content {
    width: 50rem;
    padding-right: 10px;
    padding-left: 10px;

    .spinner {
      display: flex;
      justify-content: center;
    }

    .story {
      border-color: $orange;
    }

    .comments {
      margin-bottom: 50px;

      .comment-wrapper {

        .parent {

          .btn {
            width: 130px;
            height: 38px;
            border-color: $grey;
          }
        }

        .child {
          border: none;
        }
      }
    }
  }
}
