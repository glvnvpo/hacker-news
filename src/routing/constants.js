export const ROOT_PATH                 = '/';
export const MAIN_PAGE_PATH            = '/hacker-news';
export const SINGLE_STORY_PAGE_PATH    = `${MAIN_PAGE_PATH}/:id`;