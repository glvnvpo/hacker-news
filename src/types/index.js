// @flow
export type ID = number | string;

export type Story = {
    id: ID;
    title: string;
    by: string;
    time: number | string;
    score: number | string;
    kids?: Array<number | string>;
    text?: string;
    url?: string;
    descendants?: number | string;
}

export type Comment = {
    id: ID;
    by: string;
    text: string;
    time: number | string;
    deleted?: boolean;
    dead?: boolean;
    kids?: Array<number | string>;
    children?: Array<Comment>;
    showChildComment?: boolean;
    isLoadingChildren?: boolean;
}