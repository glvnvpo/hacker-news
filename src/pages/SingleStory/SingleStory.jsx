// @flow

import React, {useState, useEffect}  from 'react';
import {useParams, useNavigate, useLocation} from "react-router-dom";
import {Button} from "react-bootstrap";
import axios from 'axios';
import {isEmpty, isNull} from "lodash";
import './styles.scss';
import type {Story, Comment, ChildComment} from '../../types';
import {ITEM} from "../../api/constants";
import {MINUTE} from "../../constants/time";
import {MAIN_PAGE_PATH} from "../../routing/constants";
import {Spinner} from "../../components/Spinner";
import {StoryCard} from "../../components/StoryCard";
import {CommentCard} from "../../components/CommentCard";

export const SingleStory = () => {
	const {id} = useParams();

	const [story, setStory] = useState({});
	const [comments, setComments] = useState([]);
	const [isStoryLoading, setStoryLoading] = useState(true);
	const [isCommentsLoading, setCommentsLoading] = useState(true);

	const navigate = useNavigate();
	let location = useLocation();

	let timer;

	useEffect(() => {
		updateStoryAndCommentsEachMinute(id);
		return () => clearTimeout(timer);
	}, [location?.pathname]);

	const loadStory = (id: number | string): Promise<Story | string> => {
		return new Promise((resolve, reject) => {
			axios(ITEM(id))
				.then(({data}) => {
					if (data) {
						setStory(data);
						resolve(data);
					}
					else {
						setStory({});
						reject('No data');
					}
				})
				.catch((err) => reject(err))
				.finally(() => setStoryLoading(false));
		});
	};

	const updateStoryAndComments = (id: number | string, event = undefined) => {
		if (event) {
			setCommentsLoading(true);
		}

		loadStory(id)
			.then((data) => loadComments(data))
			.catch(() => setCommentsLoading(false));
	};

	const updateStoryAndCommentsEachMinute = (id: number | string) => {
		updateStoryAndComments(id);
		timer = setTimeout(() => updateStoryAndCommentsEachMinute(id), MINUTE);
	};

	const goBackToStories = () => {
		navigate(MAIN_PAGE_PATH);
	};

	const loadComments = ({kids}: Story = {}) => {
		if (kids && !isEmpty(kids)) {
			const promises = kids.map(loadOneComment);

			Promise.allSettled(promises)
				.then(data => {
					const loadedComments = data
						.filter(({status})=> status === 'fulfilled')
						.map(({value}) => value);
					const childrenPromises = loadedComments.map(loadChildrenComments);
					return Promise.all(childrenPromises);
				})
				.then((loadedComments) => {
					setComments(prevComments => {
						if (!isEmpty(prevComments)) {
							return loadedComments.map(comment => {
								const existingComment = prevComments.find(el => el.id === comment.id);
								return {
									...comment,
									showChildComment: existingComment ? existingComment.showChildComment : false
								};
							});
						}
						else return loadedComments;
					});
				})
				.finally(() => setCommentsLoading(false));
		}
		else {
			setCommentsLoading(false);
		}
	};

	const loadChildrenComments = (comment: Comment): Promise<Comment> => {
		const {id, kids} = comment || {};

		return new Promise((resolve) => {
			if (id && kids && !isEmpty(kids)) {
				const promises = kids.map(loadOneComment);
				Promise.allSettled(promises)
					.then(data => {
						const loadedChildrenComments = data
							.filter(({status}) => status === 'fulfilled')
							.map(({value}) => value);

						resolve({...comment,
							children: loadedChildrenComments
						});
					});
			}
			else resolve(comment);
		});
	};

	const loadOneComment = (id: string | number): Promise<Comment | string> => {
		return new Promise((resolve, reject) => {
			axios(ITEM(id))
				.then(({data}) => !isNull(data) ? resolve(data) : reject('No data'))
				.catch(err => reject(err));
		});
	};

	const changeVisibilityOfChildComment = (parentId: number | string): Array<Comment> => {
		setComments(prevComments => {
			const parentComment = prevComments.find(({id}) => parentId === id);
			const newVisibility = !parentComment.showChildComment;
			return prevComments.map(
				comment => comment.id === parentId ?
					{...parentComment, showChildComment: newVisibility}
					: comment
			);
		});
	};

	return (
		<div className="single-story">
			<div className='content'>

				<Button className="mt-20 mb-10" onClick={() => goBackToStories()} variant="outline-primary">Go back to news</Button>

				<StoryCard isLoading={isStoryLoading} story={story} />

				{
					!isEmpty(story) &&
					<Button onClick={(e) => updateStoryAndComments(id, e)} variant="outline-success">Update comments</Button>
				}

				<div className='comments mt-20 mb-20'>

					{ (!isEmpty(story) && !isCommentsLoading) &&
					(!isEmpty(comments) ? <h6 className='color-dark-grey'>Comments:</h6>
						: <h6 className='color-grey'>No comments found</h6>)
					}

					{ (isCommentsLoading && !isStoryLoading) ? <Spinner className="mt-20" /> :
						!isEmpty(comments) && comments.map((comment: Comment = {}) =>
							<div className='comment-wrapper mt-20' key={comment.id}>
								{ comment.by &&
									<CommentCard
										comment={comment}
										onChangeVisibilityOfChildComment={changeVisibilityOfChildComment}
									/>
								}

								{ (!isEmpty(comment.children) && comment.showChildComment) && comment.children.map((childComment: ChildComment) =>
									<CommentCard
										comment={childComment}
										parent={false}
										key={childComment.id}
									/>
								)}
							</div>
						)}
				</div>
			</div>
		</div>);
};