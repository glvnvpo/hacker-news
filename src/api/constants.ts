// @flow

import {ID, URL} from '../types';

const BASE = 'https://hacker-news.firebaseio.com/v0';

export const NEW_STORIES = (): URL => `${BASE}/newstories.json`;
export const ITEM = (id: ID): URL => `${BASE}/item/${id}.json`;