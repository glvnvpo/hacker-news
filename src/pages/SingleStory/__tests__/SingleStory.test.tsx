import React, {Component} from 'react';
import {mount, ReactWrapper} from 'enzyme';
import {BrowserRouter, useNavigate} from 'react-router-dom';
import axios from 'axios';
import {isNull} from 'lodash';
import {SingleStory} from '../SingleStory';
import {Comment, Story} from '../../../types';
import {ITEM as mockITEM, NEW_STORIES as mockNEW_STORIES} from '../../../api/constants';
import {waitForComponentToPaint} from '../../../jest/helpers/wait-for-component-to-paint';

jest.mock('react-router-dom', () => ({
	...jest.requireActual('react-router-dom'),
	useNavigate: jest.fn(() => () => {
	}),
	useParams: jest.fn(() => ({id: 123}))
}));

jest.mock('lodash', () => ({
	...jest.requireActual('lodash'),
	isNull: jest.fn((value) => value === null)
}));

let mockStory: Promise<{ data: Story | null }> = Promise.resolve({
	data: {
		id: 123,
		title: 'someTitle',
		text: 'someText',
		by: 'coolUser',
		time: 1666689031,
		score: 3,
		url: 'https://reactjs.org/',
		descendants: 10,
		kids: [125]
	}
});
let mockParentComment: Promise<{ data: Comment | null }> = Promise.resolve({
	data: {
		id: 125,
		parent: 123,
		text: 'someText',
		by: 'coolUser',
		time: 1666689031,
		kids: [127]
	}
});
let mockChildComment = Promise.resolve({
	data: {
		id: 127,
		parent: 125,
		text: 'someText127',
		by: 'coolUser127',
		time: 1666689031
	}
});
jest.mock('axios', () => ({
	__esModule: true,
	...jest.requireActual('axios'),
	default: jest.fn((url) => {
		if (url === mockNEW_STORIES()) {
			return Promise.resolve({data: [10]});
		} else if (url === mockITEM(123)) {
			return mockStory;
		} else if (url === mockITEM(125)) {
			return mockParentComment;
		} else if (url === mockITEM(127)) {
			return mockChildComment;
		}
	})
}));

describe('SingleStory', () => {
	let wrapper: ReactWrapper<any, Readonly<{}>, Component<{}, {}, any>>;
	
	const getWrapper = () => {
		jest.clearAllMocks();
		return mount(<BrowserRouter>
			<SingleStory/>
		</BrowserRouter>);
	};

	beforeEach(() => {
		wrapper = getWrapper();
	});

	it('should render Single story when loading', () => {
		expect(wrapper).toMatchSnapshot();
	});

	it('should call navigate', async () => {
		await waitForComponentToPaint();
		wrapper.find('.go-back').at(1).simulate('click');
		expect(useNavigate).toHaveBeenCalled();
	});

	it('should call isNull with empty data', async () => {
		mockParentComment = Promise.resolve({data: null});
		wrapper = getWrapper();
		await waitForComponentToPaint();
		// @ts-ignore
		expect(isNull.mock.calls[0][0]).toBeNull();
	});
	
	it('should not call isNull', async () => {
		mockParentComment = Promise.reject('Something went wrong');
		wrapper = getWrapper();
		await waitForComponentToPaint();
		expect(isNull).not.toHaveBeenCalled();
	});

	it('should call axios twice due to setTimeout', () => {
		jest.useFakeTimers();
		mockStory = Promise.resolve({
			data: null
		});
		wrapper = getWrapper();
		jest.advanceTimersByTime(60000);
		expect(axios).toHaveBeenCalledTimes(2);
	});

	it('should call axios returns reject', () => {
		mockStory = Promise.reject('Something went wrong');
		wrapper = getWrapper();
		expect(axios).toHaveBeenCalledTimes(1);
	});
});