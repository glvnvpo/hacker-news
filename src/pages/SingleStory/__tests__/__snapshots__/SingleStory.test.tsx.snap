// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingleStory should render Single story when loading 1`] = `
<BrowserRouter>
  <Router
    location={
      {
        "hash": "",
        "key": "default",
        "pathname": "/",
        "search": "",
        "state": null,
      }
    }
    navigationType="POP"
    navigator={
      {
        "action": "POP",
        "createHref": [Function],
        "go": [Function],
        "listen": [Function],
        "location": {
          "hash": "",
          "key": "default",
          "pathname": "/",
          "search": "",
          "state": null,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <SingleStory>
      <div
        className="single-story"
      >
        <div
          className="content"
        >
          <Button
            active={false}
            className="go-back mt-20 mb-10"
            disabled={false}
            onClick={[Function]}
            variant="outline-primary"
          >
            <button
              className="go-back mt-20 mb-10 btn btn-outline-primary"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Go back to news
            </button>
          </Button>
          <StoryCard
            isLoading={true}
          >
            <Card
              body={false}
              className="story mt-10 mb-20"
              to=""
            >
              <div
                className="story mt-10 mb-20 card"
                to=""
              >
                <SpinnerComponent
                  className="mt-20 mb-20"
                >
                  <div
                    className="spinner"
                  >
                    <Spinner
                      animation="border"
                      className="mt-20 mb-20"
                      variant="secondary"
                    >
                      <div
                        className="mt-20 mb-20 spinner-border text-secondary"
                      />
                    </Spinner>
                  </div>
                </SpinnerComponent>
              </div>
            </Card>
          </StoryCard>
          <div
            className="comments mt-20 mb-20"
          />
        </div>
      </div>
    </SingleStory>
  </Router>
</BrowserRouter>
`;
